rbd/watchers_mimic_test.go

```diff
@@ -0,0 +1,160 @@
+// +build !luminous,!mimic
+//
+// Ceph Nautilus is the first version that supports watchers through librbd.
+
+package rbd
+
+import (
+	"testing"
+
+	"github.com/ceph/go-ceph/rados"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestListWatchers(t *testing.T) {
+	conn, _ := rados.NewConn()
+	conn.ReadDefaultConfigFile()
+	conn.Connect()
+
+	poolname := GetUUID()
+	err := conn.MakePool(poolname)
+	assert.NoError(t, err)
+
+	ioctx, err := conn.OpenIOContext(poolname)
+	assert.NoError(t, err)
+
+	name := GetUUID()
+	image, err := Create(ioctx, name, 1<<22, 22)
+	assert.NoError(t, err)
+
+	// will not work on an image that is not open
+	_, err = image.ListWatchers()
+	assert.Equal(t, ErrImageNotOpen, err)
+
+	err = image.Open(true) // readonly
+	assert.NoError(t, err)
+	watchers, err := image.ListWatchers()
+	assert.NoError(t, err)
+	assert.Equal(t, 0, len(watchers))
+
+	// testing, UpdateWatch() keeps watchers == 0
+	// open the image a 2nd time and set a watcher
+	image2 := GetImage(ioctx, name)
+	err = image2.Open()
+	assert.NoError(t, err)
+
+	_, err = image2.updateWatch()
+	assert.NoError(t, err)
+
+	watchers, err = image.ListWatchers()
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(watchers))
+
+	image3 := GetImage(ioctx, name)
+	err = image3.Open()
+	assert.NoError(t, err)
+
+	watchers, err = image.ListWatchers()
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(watchers))
+
+	watchers, err = image2.ListWatchers()
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(watchers))
+
+	watchers, err = image3.ListWatchers()
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(watchers))
+
+	// Close() also drops the watch
+	image3.Close()
+
+	watchers, err = image.ListWatchers()
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(watchers))
+
+	watchers, err = image2.ListWatchers()
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(watchers))
+
+	image2.Close()
+
+	watchers, err = image.ListWatchers()
+	assert.NoError(t, err)
+	assert.Equal(t, 0, len(watchers))
+
+	err = image.Remove()
+	assert.NoError(t, err)
+
+	ioctx.Destroy()
+	conn.DeletePool(poolname)
+	conn.Shutdown()
+}
+
+func TestUpdateUnwatch(t *testing.T) {
+	// Image.updateUnwatch() is not stable
+	t.SkipNow()
```

Comment:
This test is being skipped without even a user facing message. Should we just decline adding it at all for now? Also, what's unstable about it? The API, librados crash, something else?

Reply:
No crashes, it just does not always/immediately remove the watcher. I am not sure yet how the API is supposed to work. Will spend some time during the day to test a little more.

Reply:
Actually, I'll drop most of the test parts and private clutter. Opening an image writable adds a watcher automatically, closing the image removes it. Lets keep it simple for now, and add more functionality when there is a good way for callbacks.

Reply:
I didn't know that, that's even better!